### DISCUSSION TITLE
Testing of new bluebanquise-diskless tool

### DISCUSSION THREAD

**Discussion by ginomcevoy** [bluebanquise/bluebanquise#897] [General]
Hello,

I am opening this thread to discuss how the new bluebanquise-diskless tool can be developed and tested. We are planning to use this tool in a production environment, and we see that there are some fixes required and some improvements that can be made. 

I understand the tool is in a "beta" stage, and I am not expecting a big effort on BB side to bring it out of beta stage as soon as possible. But on our side we have tested it in a controlled environment and found ways that we can improve the tool. So there are a couple of points that want to discuss:

- What would be the best way that we can bring fixes/improvements to the tool? Should be upload incremental PRs and bump the fix version of the pxe_stack role at each time? What about the collection, do we bump the fix version too? There is one PR that has been opened last week, we have a couple more "in the pipeline".

- Do you have automated tests for the tool already available? It's hard for me at the moment to implement an automated test that goes through the workflow (bootstrap image -> reference image -> live image).

- Related to the previous question, how trustworthy are the existing images and script to generate the bootstrap images? I was not able to use the AlmaLinux bootstrap image, apparently it doesn't have sudo (required to convert it as reference image). Also, when I tried to use a script similar to the one available in the infrastructure repo (RHEL_in.sh) I was not able to boot the image. If these issues are known, I can submit a PR to improve RHEL_in.sh script.

Thanks!

**Comment by oxedions** [comment]
Hello @giacomo-mcevoy 

Sorry for the very long delay of my answer :blush: 

> What would be the best way that we can bring fixes/improvements to the tool? Should be upload incremental PRs and bump the fix version of the pxe_stack role at each time? What about the collection, do we bump the fix version too? There is one PR that has been opened last week, we have a couple more "in the pipeline".

PRs as usual, and yes, we have to increment both galaxy and role version for each changes. However, since the tool is indeed in beta stage, we can do all in a separated branch for now and merge the changes at the end. This would be convenient for everybody.
Thanks a lot for the PR, I will have a look on it asap :slightly_smiling_face: 

> Do you have automated tests for the tool already available? It's hard for me at the moment to implement an automated test that goes through the workflow (bootstrap image -> reference image -> live image).

Unfortunately no. One of the main things I need to implement in the tool now is the simple ability to be called from a command line directly instead of using the menu. This would allow to automate testings.
For you to know, all my auto tests are either the content of workflow in the main bluebanquise repository, or the Validation folder in the infrastructure repository.

> Related to the previous question, how trustworthy are the existing images and script to generate the bootstrap images? I was not able to use the AlmaLinux bootstrap image, apparently it doesn't have sudo (required to convert it as reference image). Also, when I tried to use a script similar to the one available in the infrastructure repo (RHEL_in.sh) I was not able to boot the image. If these issues are known, I can submit a PR to improve RHEL_in.sh script.

Yes, you are right, the script is broken. I made few mistakes when I tried to fix RHEL 9. I would be happy to get PRs on that too, I will try to work on it as soon as possible. Also it would be nice to provide RHEL images, but I need to check how I can pull a real RHEL image and get access to repositories since it requires a subscription.

Once DNS is ok, and PXE preseed is ok, if there are no more emergencies, I will work on this part :blush: 
At some point, I also really need to improve documentation.

**Reply by ginomcevoy** [reply]
Hello @oxedions ,
Thanks for the responses, and apologies for my late reply. I can now dedicate my time on improving the existing bluebanquise-diskless script. This should not be too hard, we have a working version that is soon to be deployed in production.

I agree on the idea of working on a branch, I think it will make pushing many changes easier. But it looks like I don't have permission to create a branch on the bluebanquise repo. I can create one at my fork, and start pushing commits there. If you create a branch for me in this repo, I can open a PR against it to review the changes.

I will push some fixes first, but I think we can also do some quality improvements on the code if you agree (some of them may be more of subjective style than others).

FYI checked the images at bluebanquise.com/diskless, managed to boot the Rocky 9.3 OK, nice!

**Reply by oxedions** [reply]
Dear @ginomcevoy 

Thanks a lot ! :blush: 

Ok ! It seems one of the most requested feature is the missing capability to use chroot as before to deploy things.
Once you start pushing commits in the branch, can you open a PR as draft for the branch, so we can see easily changes ? :slightly_smiling_face:
